src/components/tasks/TaskSearch/TaskSearch.tsx

```diff
@@ -16,84 +23,316 @@ const initialSuggestionCoordinates: SuggestionCoordinates = {
 };
 
 const TaskSearch = ({
-    onSelect,
+    onFilterDropdownSelect,
+    filterDropdownActiveTab,
     inputValue,
-    activeTab,
-    onInputChange,
     onClickSearchButton,
 }: TaskSearchProps) => {
-    const [modalOpen, setModalOpen] = useState(false);
+    const [filterDropdownModelOpen, setFilterDropdownModelOpen] =
+        useState(false);
+
+    const userInputRef = useRef<HTMLInputElement>(null);
+    const [typedInput, setTypedInput] = useState('');
+    const defferedUserInput: string = useDebounce(typedInput, 300);
+
+    const [selectedFilters, setSelectedFilters] = useState<
+        Array<TaskSearchOption>
+    >(convertStringToOptions(inputValue));
+    const [onEditSelectedFilterValue, setOnEditSelectedFilterValue] =
+        useState<string>('');
+    const [onEditSelectedFilterIndex, setOnEditSelectedFilterIndex] =
+        useState<false | number>(false);
+    const [onRemoveSelectedFilterIndex, setOnRemoveSelectedFilterIndex] =
+        useState(-1);
+    const defferedPillValue = useDebounce(onEditSelectedFilterValue, 300);
+
+    const [
+        activeFilterSuggestionDropdownIndex,
+        setActiveFilterSuggestionDropdownIndex,
+    ] = useState(-1);
+    const [suggestionCoordinates, setSuggestionCoordinates] =
+        useState<SuggestionCoordinates>(initialSuggestionCoordinates);
 
     const searchButtonHandler = () => {
-        onClickSearchButton();
+        if (onEditSelectedFilterIndex === false) {
+            onClickSearchButton(convertSearchOptionsToQuery(selectedFilters));
+        }
     };
 
+    const { filterSuggestions } = useFilterSuggestion({
+        typedInput,
+        defferedPillValue,
+        defferedUserInput,
+        onEditSelectedFilterIndex,
+        onEditSelectedFilterValue,
+        selectedFilters,
+        activeFilterSuggestionDropdownIndex,
+        setActiveFilterSuggestionDropdownIndex,
+    });
+
     const handleModal = () => {
-        setModalOpen(!modalOpen);
+        setFilterDropdownModelOpen(!filterDropdownModelOpen);
     };
+    useEffect(() => {
+        setSelectedFilters(convertStringToOptions(inputValue));
+    }, [inputValue]);
 
+    const toggleInputFocus = (inFocus = true) => {
+        inFocus && userInputRef.current?.focus();
+        !inFocus && userInputRef.current?.blur();
+    };
     const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
         switch (event.key) {
-            case 'Enter':
-                if (inputValue.length > 0) {
+            case 'Backspace':
+                if (
+                    onEditSelectedFilterIndex !== false &&
+                    onEditSelectedFilterValue.length === 1
+                ) {
+                    const newOptions = selectedFilters.filter(
+                        (_, idx) => idx !== onEditSelectedFilterIndex
+                    );
+                    setSelectedFilters(newOptions);
+                    setOnEditSelectedFilterIndex(false);
```

Comment:
`handleKeyPress` function could be refactored for better readability. It handles multiple key events and actions. Consider breaking it down into smaller functions